// minip/dist/router/Router.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Router = void 0;
class Router {
    routes = [];
    middlewares = [];
    use(middleware) {
        this.middlewares.push(middleware);
        return this;
    }
    get(path, handler) {
        return this.addRoute('GET', path, handler);
    }
    post(path, handler) {
        return this.addRoute('POST', path, handler);
    }
    put(path, handler) {
        return this.addRoute('PUT', path, handler);
    }
    delete(path, handler) {
        return this.addRoute('DELETE', path, handler);
    }
    patch(path, handler) {
        return this.addRoute('PATCH', path, handler);
    }
    async handle(req, res) {
        this.attachHelpers(res);
        const parsedUrl = new URL(req.url || '/', `http://${req.headers.host}`);
        req.query = Object.fromEntries(parsedUrl.searchParams.entries());
        const path = parsedUrl.pathname;
        const method = req.method?.toUpperCase() ?? 'GET';
        await this.runMiddlewares(req, res);
        if (res.writableEnded)
            return;
        const match = this.matchRoute(method, path);
        if (!match) {
            this.sendJson(res, 404, { error: 'Not Found' });
            return;
        }
        req.params = match.params;
        try {
            await Promise.resolve(match.route.handler(req, res));
        }
        catch (err) {
            this.sendJson(res, 500, {
                error: 'Internal Server Error',
                message: err.message,
            });
        }
    }
    addRoute(method, path, handler) {
        const { regex, paramNames } = this.pathToRegex(path);
        this.routes.push({ method, path, regex, paramNames, handler });
        return this;
    }
    pathToRegex(path) {
        const paramNames = [];
        const escape = (s) => s.replace(/[.+*?=^!:${}()[\]|/\\]/g, '\\$&');
        let regexStr = '';
        let lastIndex = 0;
        const paramPattern = /:(\w+)/g;
        let match;
        while ((match = paramPattern.exec(path)) !== null) {
            regexStr += escape(path.slice(lastIndex, match.index));
            paramNames.push(match[1]);
            regexStr += '([^/]+)';
            lastIndex = match.index + match[0].length;
        }
        regexStr += escape(path.slice(lastIndex));
        return { regex: new RegExp(`^${regexStr}$`), paramNames };
    }
    matchRoute(method, path) {
        for (const route of this.routes) {
            if (route.method !== method)
                continue;
            const matches = path.match(route.regex);
            if (!matches)
                continue;
            const params = route.paramNames.reduce((acc, name, i) => ({ ...acc, [name]: matches[i + 1] }), {});
            return { route, params };
        }
        return null;
    }
    runMiddlewares(req, res) {
        const stack = [...this.middlewares];
        const run = (index) => {
            if (index >= stack.length)
                return Promise.resolve();
            const middleware = stack[index];
            return new Promise((resolve, reject) => {
                try {
                    const result = middleware(req, res, () => {
                        run(index + 1).then(resolve).catch(reject);
                    });
                    if (result instanceof Promise) {
                        result.catch(reject);
                    }
                }
                catch (err) {
                    reject(err);
                }
            });
        };
        return run(0);
    }
    attachHelpers(res) {
        res.json = (data) => this.sendJson(res, 200, data);
        res.status = (code) => {
            res.json = (data) => this.sendJson(res, code, data);
            return res;
        };
        res.xml = (xmlString) => {
            res.writeHead(200, { 'Content-Type': 'application/xml' });
            res.end(xmlString);
        };
    }
    sendJson(res, status, data) {
        if (res.writableEnded)
            return;
        res.writeHead(status, { 'Content-Type': 'application/json' });
        res.end(JSON.stringify(data));
    }
}
exports.Router = Router;
