// minip/dist/server/Server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Server = void 0;
exports.createApp = createApp;
const http_1 = __importDefault(require("http"));
const Router_1 = require("../router/Router");
const bodyParser_1 = require("../middleware/bodyParser");
class Server {
    router;
    options;
    constructor(router, options = {}) {
        this.router = router;
        this.options = {
            port: options.port ?? 3000,
            host: options.host ?? 'localhost',
            cors: options.cors ?? '*',
        };
    }
    listen(callback) {
        const server = http_1.default.createServer(async (req, res) => {
            const allowedOrigin = Array.isArray(this.options.cors)
                ? this.options.cors.join(', ')
                : this.options.cors;
            res.setHeader('Access-Control-Allow-Origin', allowedOrigin);
            res.setHeader('Access-Control-Allow-Methods', 'GET, POST, PUT, PATCH, DELETE, OPTIONS');
            res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization');
            if (req.method === 'OPTIONS') {
                res.writeHead(204);
                res.end();
                return;
            }
            const methodsWithBody = ['POST', 'PUT', 'PATCH'];
            const contentType = req.headers['content-type'] ?? '';
            const isJson = contentType.includes('application/json');
            if (req.method && methodsWithBody.includes(req.method) && isJson) {
                try {
                    req.body = await (0, bodyParser_1.bodyParser)(req);
                }
                catch {
                    res.writeHead(400, { 'Content-Type': 'application/json' });
                    res.end(JSON.stringify({ error: 'Invalid JSON body' }));
                    return;
                }
            }
            await this.router.handle(req, res);
        });
        const { port, host } = this.options;
        server.listen(port, host, () => {
            callback?.();
            console.log(`Server running at http://${host}:${port}`);
        });
        return server;
    }
}
exports.Server = Server;
function createApp(options = {}) {
    const router = new Router_1.Router();
    const server = new Server(router, options);
    return {
        router,
        listen: (callback) => server.listen(callback),
    };
}
